Forge.TreeWalker/src/TreeWalkerSession.cs

```diff
@@ -305,31 +305,63 @@ public async Task<string> WalkTree(string treeNodeKey)
-                        this.Parameters.TreeName,
-                        this.Parameters.RootSessionId,
-                        this.walkTreeCts.Token).ConfigureAwait(false);
+                    TreeNodeContext treeNodeContext = null;
+                    if (this.Parameters.CallbacksV2 != null)
+                    {
+                        // If applicable, use the new CallbacksV2 with ITreeWalkerCallbacksV2
+                        // Evaluate the dynamic properties that are used by the actionTask.
+                        treeNodeContext = new TreeNodeContext(
+                            this.Parameters.SessionId,
+                            treeNodeKey,
+                            await this.EvaluateDynamicProperty(this.Schema.Tree[current].Properties, null),
+                            this.Parameters.UserContext,
+                            this.walkTreeCts.Token,
+                            this.Parameters.TreeName,
+                            this.Parameters.RootSessionId,
+                            shouldSkipActionsInTreeNode:false
+                        );
+
+                        // Follow the previous Callbacks with ITreeWalkerCallbacks.
+                        await this.Parameters.CallbacksV2.BeforeVisitNode(treeNodeContext).ConfigureAwait(false);
+                    }
+                    else
+                    {
+                        // Follow the previous Callbacks with ITreeWalkerCallbacks.
+                        await this.Parameters.Callbacks.BeforeVisitNode(
+                            this.Parameters.SessionId,
+                            current,
+                            await this.EvaluateDynamicProperty(this.Schema.Tree[current].Properties, null),
+                            this.Parameters.UserContext,
+                            this.Parameters.TreeName,
+                            this.Parameters.RootSessionId,
+                            this.walkTreeCts.Token).ConfigureAwait(false);
+                    }
 
                     try
                     {
                         // Exceptions are thrown here if VisitNode hit a timeout, was cancelled, or failed.
-                        next = await this.VisitNode(current).ConfigureAwait(false);
+                        bool skipAction = treeNodeContext != null && treeNodeContext.ShouldSkipActionsInTreeNode;
+                        next = await this.VisitNode(current, skipAction).ConfigureAwait(false);
                     }
                     finally
                     {
-                        // Always call this callback, whether or not visit node was successful.
-                        await this.Parameters.Callbacks.AfterVisitNode(
+                        if (treeNodeContext != null)
+                        {
+                            // Follow the previous Callbacks with ITreeWalkerCallbacks.
+                            await this.Parameters.CallbacksV2.AfterVisitNode(treeNodeContext).ConfigureAwait(false);
```

Comment:
>treeNodeContext [](start = 77, length = 15)

Hmm.. Let's discuss re-using the same TreeNodeContext object for BeforeVisitNode and AfterVisitNode. It is the same object, so it makes sense to use the same object. However, the existing behavior recalculates the Properties value. The user could potentially be relying on Properties getting calculated twice. Or they could be assuming it should only be calculated once.. Not sure which way to go here.

Reply:
Another thing to consider is that BeforeVisitNode will be manipulating property values of TreeNodeContext, such as the new SkipAction property. Given that case, we would need to pass the same object to AfterVisitNode.

---
In reply to: [504167224](https://github.com/microsoft/Forge/pull/32#discussion_r504167224) [](ancestors = 504167224)

Reply:
agree with you on both. So i now changed it to use different objects, for the recalculating, and copy the value of ShouldSkipActionsInTreeNode

---
In reply to: [504228812](https://github.com/microsoft/Forge/pull/32#discussion_r504228812) [](ancestors = 504228812,504167224)

Reply:
Ok sounds good. Doesn't look like you set currentNodeSkipActionContext on AfterVisitNode's TreeNodeContext. This should be set to the value from BeforeVistNode's TreeNodeContext.

---
In reply to: [504291737](https://github.com/microsoft/Forge/pull/32#discussion_r504291737) [](ancestors = 504291737,504228812,504167224)

Reply:
good catch. fixed

---
In reply to: [505826440](https://github.com/microsoft/Forge/pull/32#discussion_r505826440) [](ancestors = 505826440,504291737,504228812,504167224)